### ISSUE TITLE
Raw module on Windows fails to work with simple Powershell command

### ISSUE THREAD

**Issue by dagwieers** [ansible/ansible#16216]
##### ISSUE TYPE

- Bug Report
##### ANSIBLE VERSION

v2.1
##### SUMMARY

A simple Powershell command that works fine in Powershell, fails to work when run through the `raw` module.
##### STEPS TO REPRODUCE

``` yaml
- hosts: all
  tasks:
  - raw: 'Get-AppxPackage {{ item }} | Remove-AppxPackage'
    with_items:
    - Microsoft.WindowsAlarms
    - Microsoft.People
    - Microsoft.ConnectivityStore
    - Microsoft.WindowsStore
    - Microsoft.XboxApp
    - E046963F.LenovoCompanion
    - Microsoft.WindowsMaps
    - Microsoft.ZuneVideo
    - Microsoft.3DBuilder
    - Microsoft.Office.Sway
    - Microsoft.WindowsCamera
    - Microsoft.ZuneMusic
    - Microsoft.windowscommunicationsapps
    - Microsoft.CommsPhone
    - Microsoft.Windows.Photos
    - Microsoft.BingFinance
    - Microsoft.BingNews
    - Microsoft.MicrosoftSolitaireCollection
    - Microsoft.BingSports
    - Microsoft.BingWeather
```

##### EXPECTED RESULTS

Each of the apps should be removed.
##### ACTUAL RESULTS

```
Using /home/dag/home-made/crombeen/ansible.cfg as config file
SSH password: 

PLAYBOOK: remove.yml ***********************************************************
1 plays in remove.yml

PLAY [all] *********************************************************************

TASK [raw] *********************************************************************
task path: /home/dag/home-made/crombeen/remove.yml:3
<192.168.1.33> ESTABLISH WINRM CONNECTION FOR USER: ictadmin on PORT 5986 TO 192.168.1.33
<192.168.1.33> EXEC Get-AppxPackage Microsoft.WindowsAlarms | Remove-AppxPackage
ok: [windows1] => (item=Microsoft.WindowsAlarms) => {"invocation": {"module_args": {"_raw_params": "Get-AppxPackage Microsoft.WindowsAlarms | Remove-AppxPackage"}, "module_name": "raw"}, "item": "Microsoft.WindowsAlarms", "rc": 0, "stderr": "", "stdout": "", "stdout_lines": []}
<192.168.1.33> EXEC Get-AppxPackage Microsoft.People | Remove-AppxPackage
ok: [windows1] => (item=Microsoft.People) => {"invocation": {"module_args": {"_raw_params": "Get-AppxPackage Microsoft.People | Remove-AppxPackage"}, "module_name": "raw"}, "item": "Microsoft.People", "rc": 0, "stderr": "", "stdout": "", "stdout_lines": []}
<192.168.1.33> EXEC Get-AppxPackage Microsoft.ConnectivityStore | Remove-AppxPackage
ok: [windows1] => (item=Microsoft.ConnectivityStore) => {"invocation": {"module_args": {"_raw_params": "Get-AppxPackage Microsoft.ConnectivityStore | Remove-AppxPackage"}, "module_name": "raw"}, "item": "Microsoft.ConnectivityStore", "rc": 0, "stderr": "", "stdout": "", "stdout_lines": []}
<192.168.1.33> EXEC Get-AppxPackage Microsoft.WindowsStore | Remove-AppxPackage
ok: [windows1] => (item=Microsoft.WindowsStore) => {"invocation": {"module_args": {"_raw_params": "Get-AppxPackage Microsoft.WindowsStore | Remove-AppxPackage"}, "module_name": "raw"}, "item": "Microsoft.WindowsStore", "rc": 0, "stderr": "", "stdout": "", "stdout_lines": []}
<192.168.1.33> EXEC Get-AppxPackage Microsoft.XboxApp | Remove-AppxPackage
ok: [windows1] => (item=Microsoft.XboxApp) => {"invocation": {"module_args": {"_raw_params": "Get-AppxPackage Microsoft.XboxApp | Remove-AppxPackage"}, "module_name": "raw"}, "item": "Microsoft.XboxApp", "rc": 0, "stderr": "", "stdout": "", "stdout_lines": []}
<192.168.1.33> EXEC Get-AppxPackage E046963F.LenovoCompanion | Remove-AppxPackage
ok: [windows1] => (item=E046963F.LenovoCompanion) => {"invocation": {"module_args": {"_raw_params": "Get-AppxPackage E046963F.LenovoCompanion | Remove-AppxPackage"}, "module_name": "raw"}, "item": "E046963F.LenovoCompanion", "rc": 0, "stderr": "", "stdout": "", "stdout_lines": []}
<192.168.1.33> EXEC Get-AppxPackage Microsoft.WindowsMaps | Remove-AppxPackage
ok: [windows1] => (item=Microsoft.WindowsMaps) => {"invocation": {"module_args": {"_raw_params": "Get-AppxPackage Microsoft.WindowsMaps | Remove-AppxPackage"}, "module_name": "raw"}, "item": "Microsoft.WindowsMaps", "rc": 0, "stderr": "", "stdout": "", "stdout_lines": []}
<192.168.1.33> EXEC Get-AppxPackage Microsoft.ZuneVideo | Remove-AppxPackage
ok: [windows1] => (item=Microsoft.ZuneVideo) => {"invocation": {"module_args": {"_raw_params": "Get-AppxPackage Microsoft.ZuneVideo | Remove-AppxPackage"}, "module_name": "raw"}, "item": "Microsoft.ZuneVideo", "rc": 0, "stderr": "", "stdout": "", "stdout_lines": []}
<192.168.1.33> EXEC Get-AppxPackage Microsoft.3DBuilder | Remove-AppxPackage
ok: [windows1] => (item=Microsoft.3DBuilder) => {"invocation": {"module_args": {"_raw_params": "Get-AppxPackage Microsoft.3DBuilder | Remove-AppxPackage"}, "module_name": "raw"}, "item": "Microsoft.3DBuilder", "rc": 0, "stderr": "", "stdout": "", "stdout_lines": []}
<192.168.1.33> EXEC Get-AppxPackage Microsoft.Office.Sway | Remove-AppxPackage
ok: [windows1] => (item=Microsoft.Office.Sway) => {"invocation": {"module_args": {"_raw_params": "Get-AppxPackage Microsoft.Office.Sway | Remove-AppxPackage"}, "module_name": "raw"}, "item": "Microsoft.Office.Sway", "rc": 0, "stderr": "", "stdout": "", "stdout_lines": []}
<192.168.1.33> EXEC Get-AppxPackage Microsoft.WindowsCamera | Remove-AppxPackage
ok: [windows1] => (item=Microsoft.WindowsCamera) => {"invocation": {"module_args": {"_raw_params": "Get-AppxPackage Microsoft.WindowsCamera | Remove-AppxPackage"}, "module_name": "raw"}, "item": "Microsoft.WindowsCamera", "rc": 0, "stderr": "", "stdout": "", "stdout_lines": []}
<192.168.1.33> EXEC Get-AppxPackage Microsoft.ZuneMusic | Remove-AppxPackage
ok: [windows1] => (item=Microsoft.ZuneMusic) => {"invocation": {"module_args": {"_raw_params": "Get-AppxPackage Microsoft.ZuneMusic | Remove-AppxPackage"}, "module_name": "raw"}, "item": "Microsoft.ZuneMusic", "rc": 0, "stderr": "", "stdout": "", "stdout_lines": []}
<192.168.1.33> EXEC Get-AppxPackage Microsoft.windowscommunicationsapps | Remove-AppxPackage
ok: [windows1] => (item=Microsoft.windowscommunicationsapps) => {"invocation": {"module_args": {"_raw_params": "Get-AppxPackage Microsoft.windowscommunicationsapps | Remove-AppxPackage"}, "module_name": "raw"}, "item": "Microsoft.windowscommunicationsapps", "rc": 0, "stderr": "", "stdout": "", "stdout_lines": []}
<192.168.1.33> EXEC Get-AppxPackage Microsoft.CommsPhone | Remove-AppxPackage
ok: [windows1] => (item=Microsoft.CommsPhone) => {"invocation": {"module_args": {"_raw_params": "Get-AppxPackage Microsoft.CommsPhone | Remove-AppxPackage"}, "module_name": "raw"}, "item": "Microsoft.CommsPhone", "rc": 0, "stderr": "", "stdout": "", "stdout_lines": []}
<192.168.1.33> EXEC Get-AppxPackage Microsoft.Windows.Photos | Remove-AppxPackage
ok: [windows1] => (item=Microsoft.Windows.Photos) => {"invocation": {"module_args": {"_raw_params": "Get-AppxPackage Microsoft.Windows.Photos | Remove-AppxPackage"}, "module_name": "raw"}, "item": "Microsoft.Windows.Photos", "rc": 0, "stderr": "", "stdout": "", "stdout_lines": []}
<192.168.1.33> EXEC Get-AppxPackage Microsoft.BingFinance | Remove-AppxPackage
ok: [windows1] => (item=Microsoft.BingFinance) => {"invocation": {"module_args": {"_raw_params": "Get-AppxPackage Microsoft.BingFinance | Remove-AppxPackage"}, "module_name": "raw"}, "item": "Microsoft.BingFinance", "rc": 0, "stderr": "", "stdout": "", "stdout_lines": []}
<192.168.1.33> EXEC Get-AppxPackage Microsoft.BingNews | Remove-AppxPackage
ok: [windows1] => (item=Microsoft.BingNews) => {"invocation": {"module_args": {"_raw_params": "Get-AppxPackage Microsoft.BingNews | Remove-AppxPackage"}, "module_name": "raw"}, "item": "Microsoft.BingNews", "rc": 0, "stderr": "", "stdout": "", "stdout_lines": []}
<192.168.1.33> EXEC Get-AppxPackage Microsoft.MicrosoftSolitaireCollection | Remove-AppxPackage
ok: [windows1] => (item=Microsoft.MicrosoftSolitaireCollection) => {"invocation": {"module_args": {"_raw_params": "Get-AppxPackage Microsoft.MicrosoftSolitaireCollection | Remove-AppxPackage"}, "module_name": "raw"}, "item": "Microsoft.MicrosoftSolitaireCollection", "rc": 0, "stderr": "", "stdout": "", "stdout_lines": []}
<192.168.1.33> EXEC Get-AppxPackage Microsoft.BingSports | Remove-AppxPackage
ok: [windows1] => (item=Microsoft.BingSports) => {"invocation": {"module_args": {"_raw_params": "Get-AppxPackage Microsoft.BingSports | Remove-AppxPackage"}, "module_name": "raw"}, "item": "Microsoft.BingSports", "rc": 0, "stderr": "", "stdout": "", "stdout_lines": []}
<192.168.1.33> EXEC Get-AppxPackage Microsoft.BingWeather | Remove-AppxPackage
ok: [windows1] => (item=Microsoft.BingWeather) => {"invocation": {"module_args": {"_raw_params": "Get-AppxPackage Microsoft.BingWeather | Remove-AppxPackage"}, "module_name": "raw"}, "item": "Microsoft.BingWeather", "rc": 0, "stderr": "", "stdout": "", "stdout_lines": []}

PLAY RECAP *********************************************************************
windows1                   : ok=1    changed=0    unreachable=0    failed=0   

```

Apps are not removed, despite RC=0 and no output.

**Comment by nitzmahone** [comment]
Looks like a Windows bug. When I do this over basic auth, I just get success, no output, but when I do it over NTLM, Kerb, or cert auth, I get an error (the only appx packages I have can't be removed on server, so that's kinda expected). It behaves the same way doing it over `Enter-PSSession -ComputerName localhost -Authentication basic -Credential $cred -UseSSL -SessionOption $(New-PSSessionOption -SkipCACheck -SkipCNCheck)` - no output, no error... This definitely looks like "cannot fix" - though anything but Basic auth should work for you.

**Comment by dagwieers** [comment]
I wouldn't close this, because we haven't found the cause or a solution for a known problem.

**Comment by dagwieers** [comment]
Is there a document explaining cert auth with Ansible ? I am interested in pursuing this.

**Comment by nitzmahone** [comment]
@dagwieers I'm writing it as we speak (redoing the "intro to windows" page into several pages), and have planned to include samples for all supported auth methods. While we do support cert auth, it's a huge pain in the butt and not particularly useful in WinRM because of the way MS implemented it. It's just a mapping of cert thumbprint/subject to a (local only) username/password. You can't use domain users, and if you change the user's password, you have to update the mapping as well. I don't know anyone that actually uses it because of these restrictions, but I'd be curious if you actually have a use for it.

**Comment by dagwieers** [comment]
Well, if some raw commands fail with user/password authentication, the only other option I have is to use client certificates for authentication. There is no other option we have.

Can we report this issue to a Windows forum ? Does Microsoft accept issue reports somehow through the community ?

Another problem is related to the fact that WinRM doesn't seem to work with certificates generated using OpenSSL, and instead requires ADCS (Active Directory Certificate Services), see #16243. I think that is my problem ATM with client certificate authentication :-/

**Comment by nitzmahone** [comment]
Only way anything might ever happen with it is as a support incident., and even if you can actually get them to fix it, likely case is a private hotfix that you'd have to apply to all the machines. Is there some reason you can't use NTLM instead of basic (ansible_winm_transport=ntlm with pywinrm >=0.2.0)? It works fine with that...

**Comment by dagwieers** [comment]
@nitzmahone It simply does not work with ntlm either. Command returns RC 0, no other output, Appx is not removed. Does work on the Powershell command line... I was hoping client certificates would work, but somehow it feels we don't understand the problem yet...

**Comment by nitzmahone** [comment]
Aha- the difference is domain user vs not. Basic is the only authtype I run with a local user, all the others are domain users. If i do NTLM with a local user, `Get-AppXPackage` behaves the same way (RC=0, no output).

**Comment by nitzmahone** [comment]
Cert _does_ work though, which is ... odd (though the whole thing is odd).

**Comment by dagwieers** [comment]
I cannot make client certificate authentication work, a guide on how to set it up would be very useful at this point ;-)

PS Indeed, we only have local users.

**Comment by nitzmahone** [comment]
Aha again: my cert user is "Administrator"- when I do that command over NTLM or Basic as Administrator, it works too. So something hardcoded about Administrator for that command.

**Comment by dagwieers** [comment]
Aha ! We are using ictadmin, but let me test with Administrator instead...

**Comment by nitzmahone** [comment]
I've been procrastinating the cert part because I got it working (on the Windows side) ages ago, and I don't remember all the things I had to do. (plus it's pretty darn useless)

**Comment by dagwieers** [comment]
Still doesn't work for me with user Administrator. Neither ssl, nor ntlm.

In effect, it never worked for me through Ansible (it only works on the Powershell command line).

**Comment by nitzmahone** [comment]
I took apart the C# code for the cmdlets- it all eventually delegates to the WinRT PackageManager, which is a big black hole of DRM and native silliness, so noplace for mere mortals to go from there. I'm guessing it's a poorly-chosen default ACL on some object related to that ("Administrator; INTERACTIVE" or something), but since PackageManager doesn't appear to be managed code, I can't easily see which objects it's hitting. This is still pretty clearly not an Ansible issue at this point (since I can replicate it with straight WinRM on Windows), so I've got to get back to working on things that I can do something about. :) Suspect we're back to "Microsoft support incident"...
